### PR TITLE
SAMZA-1748: Standalone failure tests.

### PR DESCRIPTION
**Pull request by shanthoosh**
In the standalone model, a processor can leave and join the group at any point in time.  This processor reshuffle is referred to as rebalancing which results in task(work) redistribution amongst other available, live processors in the group.

Processor rebalancing in existing standalone integration tests(junit tests) is accomplished through clean shutdown of the processors. However, in real production scenarios, processor rebalancing is triggered through unclean shutdown and full garbage collection(GC) of the processors.

As a part of this patch to cover those scenarios, the following integration tests are added. 

1. Force killing the leader processor of the group.
2. Force killing a single follower in the group.
3. Force killing multiple followers in the group.
4. Force killing the leader and a follower in the  group.
5. Suspending and resuming the leader of the group.

Since existing standalone integration tests cover event consumption/production after the re-balancing phase, these new tests will just test the coordination. We'll iterate on this initial suite and add tests whenever necessary.

### CODE REVIEW

**Comment by shanthoosh** [comment]
**Verification:**

```bash
i=0
while [ $i -lt 50 ]; do
                i=`expr $i + 1`
                echo "Run " +$i 
                ./bin/integration-tests.sh /tmp/samza-tests/ standalone-integration-tests --nopassword >> ~/test-logs-runTests-1_26
done;
```

**Result:** 

```bash
grep -i 'passed' ~/test-logs-runTests-1_26 | wc -l 
250

grep -i 'failed' ~/test-logs-runTests-1_26 | wc -l 
0

```

File: samza-test/src/main/java/org/apache/samza/test/integration/LocalApplicationRunnerMain.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+import static org.apache.samza.runtime.ApplicationRunnerMain.STREAM_APPLICATION_CLASS_CONFIG;
+
+/**
+ * {@link ApplicationRunnerMain} was designed for deploying {@link StreamApplication} in yarn
+ * and doesn't work for in standalone.
+ *
+ * This runner class is built for standalone failure tests and not recommended for general use.
+ */
+public class LocalApplicationRunnerMain {
```
Comment (xinyuiscool) [inline samza-test/src/main/java/org/apache/samza/test/integration/LocalApplicationRunnerMain.java]: Please be specific about the application name and add test to it.

File: samza-test/src/main/java/org/apache/samza/test/integration/PassThroughStreamApplication.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.samza.config.Config;
+import org.apache.samza.operators.StreamGraph;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Acts as a pass through filter for all the events from a input stream.
+ */
+public class PassThroughStreamApplication implements StreamApplication {
```
Comment (xinyuiscool) [inline samza-test/src/main/java/org/apache/samza/test/integration/PassThroughStreamApplication.java]: Please use a real app name, like StandaloneIntegrationTestKafkaApp or something

File: samza-test/src/main/python/stream_processor.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import logging
+import os
+import time
+import zopkio.adhoc_deployer as adhoc_deployer
+from zopkio.remote_host_helper import get_ssh_client, exec_with_env
+import zopkio.runtime as runtime
+
+logger = logging.getLogger(__name__)
+
+class StreamProcessor:
```
Comment (xinyuiscool) [inline samza-test/src/main/python/stream_processor.py]: StreamProcessorLauncher

File: samza-test/src/main/python/tests/standalone_failure_tests.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+
+        event = threading.Event()
+        zk_client.watch_job_model(job_model_watch(event = event, expected_processors=processors.keys()))
+
+        event.wait(2 * GROUP_COORDINATION_TIMEOUT)
+
+        job_model = zk_client.get_latest_job_model()
+        for processor_id, deployer in processors.iteritems():
+            assert processor_id in job_model['containers'], 'Processor id: {0} does not exist in JobModel: {1}.'.format(processor_id, job_model)
+        assert leader_processor_id not in job_model['containers'], 'Leader processor: {0} exists in JobModel: {1}.'.format(leader_processor_id, job_model)
```
Comment (xinyuiscool) [inline samza-test/src/main/python/tests/standalone_failure_tests.py]: check also we have new leader id.

File: samza-test/src/main/python/tests/standalone_failure_tests.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+                break
+
+        event = threading.Event()
+        zk_client.watch_job_model(job_model_watch(event = event, expected_processors=processors.keys()))
+
+        event.wait(GROUP_COORDINATION_TIMEOUT * 2)
+
+        job_model = zk_client.get_latest_job_model()
+        for processor_id, deployer in processors.iteritems():
+            assert processor_id in job_model['containers'], 'Processor id: {0} does not exist in JobModel: {1}.'.format(processor_id, job_model)
```
Comment (xinyuiscool) [inline samza-test/src/main/python/tests/standalone_failure_tests.py]: check the leader is the same.

File: samza-test/src/main/python/tests/standalone_failure_tests.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+
+        event = threading.Event()
+        zk_client.watch_job_model(job_model_watch(event = event, expected_processors=processors.keys()))
+
+        event.wait(GROUP_COORDINATION_TIMEOUT * 2)
+
+        ## Verifications after leader was suspended.
+        job_model = zk_client.get_latest_job_model()
+        for processor_id, deployer in processors.iteritems():
+            assert processor_id in job_model['containers'], 'Processor id: {0} does not exist in containerModel: {1}.'.format(processor_id, job_model['containers'])
```
Comment (vjagadish1989) [inline samza-test/src/main/python/tests/standalone_failure_tests.py]: this asserts that all processors except the leader are on the group. should we also assert that the leader is _not_ in the group?

**Review by vjagadish1989 — COMMENTED** [review]
approved, thanks!

**Review by vjagadish1989 — APPROVED** [review]
approved. modulo previous comments.

File: samza-test/src/main/java/org/apache/samza/test/integration/LocalApplicationRunnerMain.java (hunk 1 above)
Comment (shanthoosh) [reply samza-test/src/main/java/org/apache/samza/test/integration/LocalApplicationRunnerMain.java]: Done.

File: samza-test/src/main/python/tests/standalone_failure_tests.py (hunk 4 above)
Comment (shanthoosh) [reply samza-test/src/main/python/tests/standalone_failure_tests.py]: Done.

File: samza-test/src/main/python/tests/standalone_failure_tests.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+    Launches three stream processors. Kills one follower processor. Waits till the group coordination timeout and
+    verifies that the final JobModel contains the leader processor and un-killed follower processor.
+    """
+    processors = {}
+    try:
+        __setup_zk_client()
+        __pump_messages_into_input_topic()
+        processors = __setup_processors()
+
+        leader_processor_id = zk_client.get_leader_processor_id()
```
[the remark this answers is not in the document]
Comment (shanthoosh) [reply samza-test/src/main/python/tests/standalone_failure_tests.py]: Agree. Added both assertions in all of the integration tests.

File: samza-test/src/main/python/tests/standalone_failure_tests.py (hunk 6 above)
Comment (shanthoosh) [reply samza-test/src/main/python/tests/standalone_failure_tests.py]: Done.

**Pull request closed without merging**